fix: improved handling of constructor arguments of type unknown.

wkerckho opened the pull request:
Relevant issue: Components.js related issue in the Solid Community Server https://github.com/CommunitySolidServer/CommunitySolidServer/issues/1182

# Purpose
The goal of these changes is to ensure that primitive constructor arguments preserve their data type when using `unknown` as argument type.

# Changes
I've updated the method `hasValueType` in ParameterPropertyHandlerRange to perform additional handling in case the type is a `ParameterRangeWildcard` (the type that is generated when using `unknown`).
- A private utility method `hasLiteralValueType` was added to encapsulate the handling of a literal.
- I've refactored the literal case to call this method
- The case when type is a `ParameterRangeWildcard`, now extracts the configured type of the argument and also calls `hasLiteralValueType` (when the found configured type is not string).

I've added testing for various cases of instantiations when using an unknown constructor argument in the file `instantiateFile-test`.

# Improvements
Possible improvements for this fix that can be performed by someone that is more familiar with the Components.js codebase:
- Because `hasValueType` has side-effects, it was not trivial to extract the logic that performs handling of the literals in an utility method. Hence I chose to throw an error when no handling was performed, allowing the calling code in `hasValueType` to continue processing when this occurs (instead of returning). This could be confusing, so maybe some refactoring is required.
- Testing: I've provided test cases in the form of integration tests, but ideally this functionality is also checked in `ParameterPropertyHandlerRange.test.ts` (but it was a bit daunting for me to make additions there).

rubensworks commented on lib/preprocess/parameterproperty/ParameterPropertyHandlerRange.ts:
```diff
@@ -78,6 +78,14 @@ export class ParameterPropertyHandlerRange implements IParameterPropertyHandler
+        // Get datatype of the configured value
+        const configuredDataType = (<any> value.term).datatype;
+        // If this datatype is not string, handle as Literal of this type
+        if (configuredDataType !== IRIS_XSD.string) {
```
Is this if-statement necessary? Can we not also just let it run for strings?

rubensworks reviewed (changes requested):
Thanks for the PR!

> Testing: I've provided test cases in the form of integration tests, but ideally this functionality is also checked in ParameterPropertyHandlerRange.test.ts (but it was a bit daunting for me to make additions there).

Indeed, we'll need some tests here as well.

I think we'll need a new test block to some cases of the new method, such as this one: https://github.com/LinkedSoftwareDependencies/Components.js/blob/master/test/unit/preprocess/parameterproperty/ParameterPropertyHandlerRange-test.ts#L82
Happy to give additional pointers if needed.

wkerckho commented:
Incorporated the requested changes and updated `ParameterPropertyHandlerRange-test.ts`:
- Moved literal test cases for `hasValueType` into new section which specifically tests `interpretValueAsType`
- Added two new test cases for `hasValueType`, which checks if `interpretValueAsType` is called the appropriate number of times, with the expected arguments (for literal types).

rubensworks commented on test/unit/preprocess/parameterproperty/ParameterPropertyHandlerRange-test.ts:
```diff
@@ -23,17 +24,34 @@ function expectOutputProperties(output: Resource | undefined, expected: Resource
 }
 
+// Copy of internal interface declaration in ParameterPropertyHandlerRange
+interface ILiteralTypeResult {
```
Can't we just import it from the ParameterPropertyHandlerRange file?

wkerckho replied:
Yes, if you're ok with exporting this interface!

rubensworks replied:
Sure! We might want to rename it slightly though, to make more sense in a global context.

rubensworks replied:
e.g. `ILiteralAsTypeInterpretationResult`. But open for other suggestions.

wkerckho replied:
Check, sounds good to me!

rubensworks commented:
Thanks for this! Released as 5.2.0.
